grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcExceptionHandlerFunctionUtil.java

```diff
@@ -13,37 +13,48 @@
  * License for the specific language governing permissions and limitations
  * under the License.
  */
-
 package com.linecorp.armeria.internal.common.grpc;
 
 import static java.util.Objects.requireNonNull;
 
 import com.linecorp.armeria.common.RequestContext;
-import com.linecorp.armeria.common.annotation.Nullable;
 import com.linecorp.armeria.common.grpc.GrpcExceptionHandlerFunction;
 import com.linecorp.armeria.common.grpc.protocol.ArmeriaStatusException;
 import com.linecorp.armeria.common.util.Exceptions;
 
 import io.grpc.Metadata;
 import io.grpc.Status;
 
-public final class UnwrappingGrpcExceptionHandleFunction implements GrpcExceptionHandlerFunction {
-    private final GrpcExceptionHandlerFunction delegate;
+public final class GrpcExceptionHandlerFunctionUtil {
 
-    public UnwrappingGrpcExceptionHandleFunction(GrpcExceptionHandlerFunction handlerFunction) {
-        delegate = handlerFunction;
+    public static Metadata generateMetadataFromThrowable(Throwable exception) {
+        final Metadata metadata = Status.trailersFromThrowable(exception);
+        return metadata != null ? metadata : new Metadata();
     }
 
-    @Nullable
-    @Override
-    public Status apply(RequestContext ctx, @Nullable Status status, Throwable cause, Metadata metadata) {
-        final Throwable t = peelAndUnwrap(cause);
-        return delegate.apply(ctx, status, t, metadata);
+    public static Status fromThrowable(RequestContext ctx, GrpcExceptionHandlerFunction exceptionHandler,
+                                       Throwable t, Metadata metadata) {
+        final Status status = Status.fromThrowable(t);
+        final Throwable cause = status.getCause();
+        if (cause == null) {
+            return status;
+        }
+        return applyExceptionHandler(ctx, exceptionHandler, status, cause, metadata);
+    }
+
+    public static Status applyExceptionHandler(RequestContext ctx,
+                                               GrpcExceptionHandlerFunction exceptionHandler,
+                                               Status status, Throwable cause, Metadata metadata) {
+        final Throwable peeled = peelAndUnwrap(cause);
```

Comment:
Should we apply `peelAndUnwrap(cause)` before calling `Status.fromThrowable(t)` because `t` could be `CompletionException(StatusExcetion())`

Reply:
Ahh, both are `public` methods. Then we may apply `peelAndUnwrap()` to both methods although it can be called twice.

Reply:
One concern is that the upstream doesn't call `peelAndUnwrap` before `Status.fromThrowable(t)`, but there's likely no harm in doing so.

Reply:
The upstream doesn't use `CompletableFuture` by which exceptions are easily wrapped with `CompletionException`.

What do you think of applying `peelAndUnwrap` also for `generateMetadataFromThrowable()`? I missed that point before.

Reply:
Addressed. 😉